common/src/main/kotlin/streams/service/sink/strategy/CUDIngestionStrategy.kt

```diff
@@ -0,0 +1,263 @@
+package streams.service.sink.strategy
+
+import streams.events.EntityType
+import streams.extensions.quote
+import streams.serialization.JSONUtils
+import streams.service.StreamsSinkEntity
+import streams.utils.IngestionUtils.getLabelsAsString
+import streams.utils.IngestionUtils.getNodeKeysAsString
+import streams.utils.StreamsUtils
+
+
+enum class CUDOperations { create, merge, update, delete }
+
+abstract class CUD {
+    abstract val op: CUDOperations
+    abstract val type: EntityType
+    abstract val properties: Map<String, Any?>
+}
+
+data class CUDNode(override val op: CUDOperations,
+                   override val properties: Map<String, Any?> = emptyMap(),
+                   val ids: Map<String, Any?> = emptyMap(),
+                   val detach: Boolean = true,
+                   val labels: List<String> = emptyList()): CUD() {
+    override val type = EntityType.node
+    
+    fun toMap(): Map<String, Any> {
+        return when (op) {
+            CUDOperations.delete -> mapOf("ids" to ids)
+            else -> mapOf("ids" to ids, "properties" to properties)
+        }
+    }
+}
+
+data class CUDNodeRel(val ids: Map<String, Any?> = emptyMap(),
+                      val labels: List<String>)
+data class CUDRelationship(override val op: CUDOperations,
+                           override val properties: Map<String, Any?> = emptyMap(),
+                           val rel_type: String,
+                           val from: CUDNodeRel,
+                           val to: CUDNodeRel): CUD() {
+    override val type = EntityType.relationship
+
+    fun toMap(): Map<String, Any> {
+        val from = mapOf("ids" to from.ids)
+        val to = mapOf("ids" to to.ids)
+        return when (op) {
+            CUDOperations.delete -> mapOf("from" to from,
+                    "to" to to)
+            else -> mapOf("from" to from,
+                    "to" to to,
+                    "properties" to properties)
+        }
+    }
+}
+
+
+class CUDIngestionStrategy: IngestionStrategy {
+
+    companion object {
+        @JvmStatic val ID_KEY = "ids"
+    }
+
+    data class NodeRelMetadata(val labels: List<String>, val ids: Set<String>)
+
+    private fun buildNodeLookupByIds(keyword: String = "MATCH", ids: Set<String>, labels: List<String>, identifier: String = "n", field: String = ""): String {
+        val fullField = if (field.isNotBlank()) "$field." else field
+        return when (ids.contains("_id")) {
+            true -> "MATCH ($identifier:${getLabelsAsString(labels)}) WHERE id($identifier) = event.$fullField$ID_KEY._id"
```

Comment:
you have to escape all identifiers and basically ALL inputs to this function with backticks.   If I pass an identifier with a space in it, or a special character, this will break.

Double check code in getLabelsAsString for the same.

Reply:
Also Lju brings up the point here -- what if labels = []?  You don't have to have labels if you're specifying an internal ID -- and this string concat may go wrong if labels is an empty array

Reply:
> you have to escape all identifiers and basically ALL inputs to this function with backticks. If I pass an identifier with a space in it, or a special character, this will break.
> 
> Double check code in getLabelsAsString for the same.

The `getLabelsAsString` method already escapes with the backticks the strings if necessary, I'll add a quote for identifiers

Reply:
> Also Lju brings up the point here -- what if labels = []? You don't have to have labels if you're specifying an internal ID -- and this string concat may go wrong if labels is an empty array

Ok for this I think we should elaborate a little bit, in case of:
- Create: do we want to allow to create a node without a label?
- UPDATE/MERGE/DELETE: as Lju pointed out if we refer to the internal ID use the labels is an overhead, I'll remove this; but in other cases?

Reply:
@conker84 labels are optional in Neo4j, so if we **don't** allow creating a node without label, we're going to get complaints and bug reports about this as a strange limitation of the sink.  So if it's not excessively complicated my view would be to allow it, and to print a warning or a suggestion that users probably shouldn't do this....we can also put in the docs that it's a bad idea.

To do the MATCH/MERGE, since indexes are always label-dependent, it's also going to be a terrible idea from a performance perspective.  Another possibility is to, with documentation, always turn `labels: []` into `labels: ["_Unlabeled"]` so that we can segregate and track these nodes, and always get a label index to be used.  In other words, for the labels array to have a default which is less than terrible -- and a warning to the user.

Reply:
Yep no problem at all, I was just brainstorming :)

Reply:
@moxious Let's manage `labels: []` and add an `N.B.` about the performance issues into the field description wdyt?